### PR TITLE
Option to put a newline to be put before the parameters of a lambda

### PR DESCRIPTION
**Pull request by stefanobaghino**
Addresses #438

### CODE REVIEW

**Comment by stefanobaghino** [comment]
`[error] (*:update) sbt.ResolveException: unresolved dependency: pl.project13.scala#sbt-jmh;0.2.6: not found`

:anguished:

**Comment by stefanobaghino** [comment]
My bad, looks like I had to rebase.

File: core/src/main/scala/org/scalafmt/internal/Router.scala [hunk 1]
```diff
@@ -173,10 +173,17 @@ class Router(formatOps: FormatOps) {
...
           startsLambda || newlines > 0
 
+        val newlineBeforeLambdaParameters =
+          formatOps.style.newlineBeforeLambdaParameters
+
         Seq(
           Split(Space, 0, ignoreIf = skipSingleLineBlock)
             .withOptimalToken(close, killOnFail = true)
             .withPolicy(SingleLineBlock(close)),
+          Split(nl, 0, ignoreIf = !newlineBeforeLambdaParameters)
```
Comment (olafurpg) [inline core/src/main/scala/org/scalafmt/internal/Router.scala]: I don't think we need a new branch here, it is probably enough to do something similar to what @melrief does in https://github.com/olafurpg/scalafmt/pull/440/files#diff-ded7f78a4a0d490270763a26fa35f6b7R828

Basically, we want to ignore the Space on line 180 if `newlines > 0`.

File: core/src/test/resources/newlineBeforeLambdaParams/LambdaNewline.stat [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+lst.map { x =>
+      println(x)
+  x + 1
+}
+>>>
+lst.map {
+  x =>
+    println(x)
+    x + 1
+}
```
Comment (olafurpg) [inline core/src/test/resources/newlineBeforeLambdaParams/LambdaNewline.stat]: What happens for this?

```
lst.map { x => x }
```

Comment (stefanobaghino) [inline core/src/test/resources/newlineBeforeLambdaParams/LambdaNewline.stat]: In this case in particular I'd like one-liners to be preserved, unless explicitly indented... but I'm struggling with it.

**Comment by stefanobaghino** [comment]
Ok, I think this can be satisfactory as a start, let me know what you think.

File: core/src/main/scala/org/scalafmt/internal/Router.scala [hunk 3]
```diff
@@ -173,11 +173,14 @@ class Router(formatOps: FormatOps) {
         val skipSingleLineBlock =
           startsLambda || newlines > 0
 
+        val lambdaParamsNl =
+          formatOps.style.newlineBeforeLambdaParameters
```
Comment (olafurpg) [inline core/src/main/scala/org/scalafmt/internal/Router.scala]: `&& newlines > 0` should do the trick to support single line lambdas

Comment (olafurpg) [inline core/src/main/scala/org/scalafmt/internal/Router.scala]: Also, you can use `style` instead of `formatOps.style`.

Comment (stefanobaghino) [inline core/src/main/scala/org/scalafmt/internal/Router.scala]: Thanks! I'll try it out.

**Comment by stefanobaghino** [comment]
Ok, I've ditched the one-liner preservation thing completely since it's not what's currently done anyway.

**Comment by olafurpg** [comment]
LGTM, thanks 😄

**Comment by olafurpg** [comment]
@stefanobaghino FYI, I renamed the flag to `newlines.alwaysBeforeCurlyBraceLambdaParams` to be consisted with a group of other `newlines.*` options.

**Comment by stefanobaghino** [comment]
@olafurpg Great, thanks for accepting the PR, can't wait to try out the new version! :tada:

File: core/src/test/resources/newlineBeforeLambdaParams/LambdaNewline.stat (hunk 2 above)
Comment (fmv1992) [reply core/src/test/resources/newlineBeforeLambdaParams/LambdaNewline.stat]: hey, I think its an overkill to open an issue for this

but how can I avoid

```
def f(x) = {
    x => x
}
```

becoming

```
def f(x) = {
    x =>
      x
}
```

?

**Pull request merged**